src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Date.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Text.Json
+{
+    internal static partial class JsonWriterHelper
+    {
+        //
+        // Trims roundtrippable DateTime(Offset) input.
+        // If the milliseconds part of the date is zero, we omit the fraction part of the date,
+        // else we write the fraction up to 7 decimal places with no trailing zeros. i.e. the output format is
+        // YYYY-MM-DDThh:mm:ss[.s]TZD where TZD = Z or +-hh:mm.
+        // e.g.
+        //   ---------------------------------
+        //   2017-06-12T05:30:45.768-07:00
+        //   2017-06-12T05:30:45.00768Z           (Z is short for "+00:00" but also distinguishes DateTimeKind.Utc from DateTimeKind.Local)
+        //   2017-06-12T05:30:45                  (interpreted as local time wrt to current time zone)
+        public static void TrimDateTimeOffsetString(Span<byte> buffer, out int bytesWritten)
+        {
+            Debug.Assert(buffer.Length == JsonConstants.MaximumFormatDateTimeLength ||
```

Comment:
I've added a comment:
```C#
// Assert buffer is the right length for:
// YYYY-MM-DDThh:mm:ss.fffffff (JsonConstants.MaximumFormatDateTimeLength)
// YYYY-MM-DDThh:mm:ss.fffffffZ (JsonConstants.MaximumFormatDateTimeLength + 1)
// YYYY-MM-DDThh:mm:ss.fffffff(+|-)hh:mm (JsonConstants.MaximumFormatDateTimeOffsetLength)
```

Reply:
When parsing a `DateTime` will the length ever be > `JsonConstants.MaximumFormatDateTimeLength`?

Are the other two cases only for `DateTimeOffset`?

Reply:
> When parsing a `DateTime` will the length ever be > `JsonConstants.MaximumFormatDateTimeLength`?

Yes, it could be. See:

```C#
using System;
using System.Buffers;
using System.Buffers.Text;
using System.Diagnostics;
using System.Text;

namespace test
{
    class Program
    {
        static void Main(string[] args)
        {
            Span<byte> mySpan = stackalloc byte[33];
            DateTime[] dts =  {
                new DateTime(2012, 04, 16, 22, 22, 22),
                new DateTime(2012, 04, 16, 22, 22, 22, DateTimeKind.Unspecified),
                new DateTime(2012, 04, 16, 22, 22, 22, DateTimeKind.Local),
                new DateTime(2012, 04, 16, 22, 22, 22, DateTimeKind.Utc),
            }; 

            foreach (DateTime dt in dts)
            {
                bool result = Utf8Formatter.TryFormat(dt, mySpan, out int bytesWritten, new StandardFormat('O'));
                Debug.Assert(result);
            
                Console.WriteLine(new string(Encoding.UTF8.GetChars(mySpan.Slice(0, bytesWritten).ToArray())));
            }
        }
    }
}
```

Output:
```
2012-04-16T22:22:22.0000000
2012-04-16T22:22:22.0000000
2012-04-16T22:22:22.0000000-04:00
2012-04-16T22:22:22.0000000Z
```

The difference between `DateTime` and `DateTimeOffset` for trimming is that a `DT` _might_ have an offset while `DTO` always will.

i.e `if (buffer.Length > JsonConstants.MaximumFormatDateTimeLength)` is always `true` for DTO), but not always for DT.